[SPARK-8389][Streaming][PySpark] Expose KafkaRDDs offsetRange in Python

jerryshao opened the pull request:
This PR propose a simple way to expose OffsetRange in Python code, also the usage of offsetRanges is similar to Scala/Java way, here in Python we could get OffsetRange like:

```
dstream.foreachRDD(lambda r: KafkaUtils.offsetRanges(r))
```

Reason I didn't follow the way what SPARK-8389 suggested is that: Python Kafka API has one more step to decode the message compared to Scala/Java, Which makes Python API return a transformed RDD/DStream, not directly wrapped so-called JavaKafkaRDD, so it is hard to backtrack to the original RDD to get the offsetRange.

jerryshao committed:
Add HasOffsetRanges for Python

jerryshao committed:
Style fix

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36378 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36378/consoleFull) for   PR 7185 at commit [`2aabf9e`](https://github.com/apache/spark/commit/2aabf9e1b7b8414aab3428cbbab772504b66078f).

tdas commented:
@koeninger Mind taking a look?

SparkQA commented:
[Test build #36378 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36378/console) for   PR 7185 at commit [`2aabf9e`](https://github.com/apache/spark/commit/2aabf9e1b7b8414aab3428cbbab772504b66078f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnHex(child: Expression) extends UnaryExpression with Serializable `

AmplabJenkins commented:
Merged build finished. Test PASSed.

jerryshao commented:
Why only running the Kafka stream part, is this test mechanism changed?

koeninger commented on external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala:
```diff
@@ -670,4 +670,9 @@ private class KafkaUtilsPythonHelper {
+
+  def getParentKafkaRDDs(rdd: RDD[_]): JList[RDD[_]] = {
+    val parentRDDs = rdd.getNarrowAncestors
+    parentRDDs.filter(rdd => rdd.isInstanceOf[KafkaRDD[_, _, _, _, _]])
```
The general approach seems good, but I don't understand why this is searching the whole list of parent rdds.  Shouldn't the original kafka rdd be a fixed distance away in the chain of parents?

This feels wrong to me for several reasons:

- There should be at most 1 parent kafka rdd, not a list of them.
- If someone calls KafkaUtils.offsetRanges on something that isn't a KafkaRDD, they're going to get an empty list.  I personally would expect an exception.
- The scala / java approach we recommend (typecast) is only going to work if it's done on the original rdd, not after a transformation.  I personally think this is a little clearer, as it reinforces the idea that you can't rely on 1:1 after a repartition.  But in any case I feel like the approach we're recommending for the different languages should be similar; either change this, or add a java / scala static method that grabs offset ranges in a similar manner.

tdas commented:
I agree with most of @koeninger points. I think the issue is that there is that the python KafkaDirectStream generates python RDDs which is effectively [python map on the JavaPairRDD wrapping the KafkaRDD].

How about this. 

* Create a KafkaRDD class in python which is a subclass of python's RDD class. It adds the method getOffsetRanges(), which will find the offset ranges by traversing through the from the Scala RDD inside associated JavaPairRDD.

* Create a KafkaDirectDStream in python which is a subclass of python's DStream, and override transform and foreachRDD method to apply the user specified function on the python's KafkaRDD objects instead of RDD objects. This will ensure that the user will be able to call `rdd.getOffsetRanges()` inside the transform and foreach functions.

With this I think it is doable to maintain consistency with Scala API, that is, `rdd.getOffsetRanges()` in python gives the offset ranges.

koeninger commented:
I don't have a problem with the static method, especially if it prevents
yet more wrapper classes.

My concern was more about the difference in semantics, not a static vs
member method.
On Jul 2, 2015 3:15 PM, "Tathagata Das" <notifications@github.com> wrote:

> I agree with most of @koeninger <https://github.com/koeninger> points. I
> think the issue is that there is that the python KafkaDirectStream
> generates python RDDs which is effectively [python map on the JavaPairRDD
> wrapping the KafkaRDD].
>
> How about this.
>
>    -
>
>    Create a KafkaRDD class in python which is a subclass of python's RDD
>    class. It adds the method getOffsetRanges(), which will find the offset
>    ranges by traversing through the from the Scala RDD inside associated
>    JavaPairRDD.
>    -
>
>    Create a KafkaDirectDStream in python which is a subclass of python's
>    DStream, and override transform and foreachRDD method to apply the user
>    specified function on the python's KafkaRDD objects instead of RDD objects.
>    This will ensure that the user will be able to call
>    rdd.getOffsetRanges() inside the transform and foreach functions.
>
> With this I think it is doable to maintain consistency with Scala API,
> that is, rdd.getOffsetRanges() in python gives the offset ranges.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/7185#issuecomment-118153576>.
>

jerryshao commented:
Hi @koeninger and @tdas , thanks a lot for your comments, I think TD already explained well why I need to use `getNarrowAncestors()` to get the original KafkaRDD, Python code not only wraps the original KafkaRDD, but also do some additional transformations, so the RDD/DStream user get is not yet original KafkaRDD/DirectKafkaInputDStream.

Also in the current implementation, the distance to the original KafkaRDD is fixed, but it really depends on the internal of PySpark/ PySpark Streaming, if the logic changed, the distance will also be changed, so we should not assume a known and fixed distance.

Yes, currently my implementation definitely have some semantic difference compared to Java/Scala, also not so strict enough.

@tdas , I tried your way, since the solution you mentioned offers the same functionalities but with more python wrappers, I chose the previous simple way what I implemented. But thought bit on your way, I think your way is more strict and similar to the Scala/Java way, and will not lead to error use for users. I will try again your way.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

jerryshao commented:
Hi @tdas , now I change the code to follow your suggestion, would you please help to review, thanks a lot.

Now we offer similar way with Scala/Java to expose offsetRanges, also minimize the error use possibility.

```
rdd.offsetRanges()

or

stream.foreachRDD(lambda rdd: rdd.offsetRanges())
```

To support this, add lots of Kafka specific wrappers, please help to review, greatly appreciate it.

jerryshao committed:
refactor the code according to the comments

SparkQA commented:
[Test build #36491 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36491/consoleFull) for   PR 7185 at commit [`cff3893`](https://github.com/apache/spark/commit/cff3893c34c7d493340fd53c988048cfe2ddbc6d).

SparkQA commented:
[Test build #36491 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36491/console) for   PR 7185 at commit [`cff3893`](https://github.com/apache/spark/commit/cff3893c34c7d493340fd53c988048cfe2ddbc6d).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KafkaRDD(RDD):`
  * `class KafkaDStream(DStream):`
  * `class KafkaTransformFunction(TransformFunction):`
  * `class KafkaTransformedDStream(TransformedDStream):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

jerryshao commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36493 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36493/consoleFull) for   PR 7185 at commit [`7debf1c`](https://github.com/apache/spark/commit/7debf1ce521ac7701f8d5fe6906ea8178fbeefcf).

jerryshao committed:
bug fix

SparkQA commented:
[Test build #36493 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36493/console) for   PR 7185 at commit [`7debf1c`](https://github.com/apache/spark/commit/7debf1ce521ac7701f8d5fe6906ea8178fbeefcf).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KafkaRDD(RDD):`
  * `class KafkaDStream(DStream):`
  * `class KafkaTransformFunction(TransformFunction):`
  * `class KafkaTransformedDStream(TransformedDStream):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

jerryshao commented:
Still has serialization bug in the code, will fix it soon.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36555 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36555/consoleFull) for   PR 7185 at commit [`fd13937`](https://github.com/apache/spark/commit/fd13937ec28f0f01e410c5247859fde1f15b6a61).

jerryshao committed:
Fix serialization bug

SparkQA commented:
[Test build #36555 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36555/console) for   PR 7185 at commit [`fd13937`](https://github.com/apache/spark/commit/fd13937ec28f0f01e410c5247859fde1f15b6a61).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KafkaRDD(RDD):`
  * `class KafkaDStream(DStream):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

tdas commented on external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala:
```diff
@@ -670,4 +670,17 @@ private class KafkaUtilsPythonHelper {
 
   def createBroker(host: String, port: JInt): Broker = Broker(host, port)
+
+  def offsetRangesOfKafkaRDD(rdd: RDD[_]): JList[OffsetRange] = {
```
This should not be a public API. Please add this method to the `KafkaUtilsPythonHelper` class.

jerryshao replied:
Hi @tdas , I think I already put this method into `KafkaUtilsPythonHelper`, not sure what you specified.

tdas replied:
Aaaahh, my bad. I didnt notice.

tdas commented on python/pyspark/streaming/dstream.py:
```diff
@@ -602,13 +602,17 @@ class TransformedDStream(DStream):
     one transformation.
     """
-    def __init__(self, prev, func):
+    def __init__(self, prev, func, transformFunc=None):
```
Why is this a parameter in the DStream constructor? Seems pretty weird to suddenly have one of the operations be randomly configurable through a field in the constructor. Why not override it in a subclass?

tdas replied:
Actually, I misunderstood this. And I am not really getting why this is necessary.

jerryshao replied:
Previously I use override to do it, but have several problem with `TransformFunction`/`TransformDStream` deserialization problem. I guess it is the derived class which makes deserialization error, so instead of using `KafkaTransformFunction`/`KafkaTransformDStream`, I added one another field to workaround this problem. I admitted this implementation is weird, do you have any better suggestion?

tdas replied:
I dont. @davies could you please take a look?
Also, could you elaborate on what were the deser errors you were seeing. Under what conditions? How to reproduce?

jerryshao replied:
Hi @tdas , my previous way like this (subclass way):

```python
class TransformedDStream(DStream):

.....

    def transfromFunc(self, ctx, func, *deserializers):
        return TransformFunction(ctx, func, deserializers)
```

```python

class KafkaTransformedDStream(TransformedDStream):

    def __init__(self, prev, func):
        TransformedDStream.__init__(self, prev, func)

    def transfromFunc(self, ctx, func, *deserializers):
        return TransformFunction(ctx, func, deserializers) \
            .rdd_wrapper(lambda jrdd, ctx, ser: KafkaRDD(jrdd, ctx, ser))
```

this will lead to such exception:

```
Traceback (most recent call last):
  File "/mnt/data/project/apache-spark/python/pyspark/streaming/util.py", line 62, in call
    r = self.func(t, *rdds)
  File "/mnt/data/project/apache-spark/python/pyspark/streaming/tests.py", line 99, in get_output
    r = rdd.collect()
  File "/mnt/data/project/apache-spark/python/pyspark/rdd.py", line 758, in collect
    port = self.ctx._jvm.PythonRDD.collectAndServe(self._jrdd.rdd())
  File "/mnt/data/project/apache-spark/python/lib/py4j-0.8.2.1-src.zip/py4j/java_gateway.py", line 538, in __call__
    self.target_id, self.name)
  File "/mnt/data/project/apache-spark/python/lib/py4j-0.8.2.1-src.zip/py4j/protocol.py", line 300, in get_return_value
    format(target_id, '.', name), value)
Py4JJavaError: An error occurred while calling z:org.apache.spark.api.python.PythonRDD.collectAndServe.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 44.0 failed 1 times, most recent failure: Lost task 0.0 in stage 44.0 (TID 44, localhost): org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/mnt/data/project/apache-spark/python/lib/pyspark.zip/pyspark/worker.py", line 111, in main
    process()
  File "/mnt/data/project/apache-spark/python/lib/pyspark.zip/pyspark/worker.py", line 105, in process
    iterator = deserializer.load_stream(infile)
AttributeError: 'tuple' object has no attribute 'load_stream'

	at org.apache.spark.api.python.PythonRDD$$anon$1.read(PythonRDD.scala:138)
	at org.apache.spark.api.python.PythonRDD$$anon$1.<init>(PythonRDD.scala:179)
	at org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:97)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:277)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:244)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:63)
	at org.apache.spark.scheduler.Task.run(Task.scala:70)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:213)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

I guess this is due to serialization issue, so I workaround this as what I did currently.

I will do some other tests to see if there's any other way.

tdas commented on python/pyspark/streaming/kafka.py:
```diff
@@ -200,14 +203,30 @@ def __init__(self, topic, partition, fromOffset, untilOffset):
+        self.fromOffset = fromOffset
+        self.untilOffset = untilOffset
+
+    def __eq__(self, other):
```
clarification: Is this necessary for this PR or is it just to make sure that equality of this python class works in the same way as the Java counterpart?

jerryshao replied:
I think it is easy to compare the equality of two objects, especially for codes in unit tests, otherwise we have to compare two `OffsetRange` filed by field.

jerryshao replied:
Also the behavior is the same as Java counterpart.

tdas replied:
Yeah thats cool.

tdas commented on python/pyspark/streaming/kafka.py:
```diff
@@ -244,3 +263,71 @@ def __init__(self, host, port):
+                KafkaUtils._printErrorMsg(self.ctx)
+            raise e
+
+        offsets = [OffsetRange(o.topic(), o.partition(), o.fromOffset(), o.untilOffset())
```
offsets --> ranges

tdas commented:
Other than a few easy-to-fix change, the general approach LGTM. 
@koeninger Any thoughts about this PR?

koeninger commented on external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala:
```diff
@@ -670,4 +670,17 @@ private class KafkaUtilsPythonHelper {
+
+  def offsetRangesOfKafkaRDD(rdd: RDD[_]): JList[OffsetRange] = {
+    val parentRDDs = rdd.getNarrowAncestors
+    val kafkaRDDs = parentRDDs.filter(rdd => rdd.isInstanceOf[KafkaRDD[_, _, _, _, _]])
```
This should probably be find, not filter.  That would eliminate the length check below.

jerryshao replied:
Personally I'd incline to use filter with `require(...)` as mentioned by TD below to guard against some unexpected situations (more than one KafkaRDDs in the lineage chain), since the lineage should not be long (2 or 3 dependencies), so this is not a big problem.

tdas replied:
Yes, I agree. filter + require guard against unexpected situations better.

koeninger commented:
In general, I feel like this is solving the problem of too many wrappers by adding more wrappers.  I don't know that it's worth it just to get an instance method instead of a static method.

That being said, as long as the semantics are similar, I'm not against it.

tdas commented:
This wrapper problem has always been a problem with the Python API. All these wrapper magic allow us to the provide the almost similar semantics and API in Python as that in Java, which is what is the ultimate goal. This wrapper magic comes with the wrapper hell, which we have considered as a necessary, unavoidable evil in order to achieve the final goal.

tdas commented on external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala:
```diff
@@ -670,4 +670,17 @@ private class KafkaUtilsPythonHelper {
+    if (kafkaRDDs.isEmpty) {
+      throw new IllegalStateException("Get offsetRanges from a non KafkaRDD is illegal")
+    } else if (kafkaRDDs.length > 1) {
+      throw new IllegalStateException("Get offsetRanges from multiple KafkaRDDs is unexpected")
```
You can collapse all of these conditions into a require
```
require(
   kafkaRDDs.length == 1, 
   "Cannot get offset ranges, as there may be multiple Kafka RDD associated with this RDD. Please call this method only on a KafkaRDD.")
```

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36631 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36631/consoleFull) for   PR 7185 at commit [`e6a8011`](https://github.com/apache/spark/commit/e6a8011b2904823f463b1b5fddec0c4ddd57ac4d).

jerryshao committed:
Address the comments

SparkQA commented:
[Test build #36631 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36631/console) for   PR 7185 at commit [`e6a8011`](https://github.com/apache/spark/commit/e6a8011b2904823f463b1b5fddec0c4ddd57ac4d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KafkaRDD(RDD):`
  * `class KafkaDStream(DStream):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36634 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36634/consoleFull) for   PR 7185 at commit [`4c6d320`](https://github.com/apache/spark/commit/4c6d3204d55b23a088f6da59028d5edeb1e7f66f).

jerryshao committed:
Another way to fix subclass deserialization issue

SparkQA commented:
[Test build #36634 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36634/console) for   PR 7185 at commit [`4c6d320`](https://github.com/apache/spark/commit/4c6d3204d55b23a088f6da59028d5edeb1e7f66f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KafkaRDD(RDD):`
  * `class KafkaDStream(DStream):`
  * `class KafkaTransformedDStream(TransformedDStream):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

jerryshao commented on python/pyspark/streaming/kafka.py:
```diff
@@ -244,3 +263,87 @@ def __init__(self, host, port):
+        return KafkaTransformedDStream(self, func)
+
+
+class KafkaTransformedDStream(TransformedDStream):
```
Hi @tdas , here is another way using override to get Kafka specific `TransformedDStream` and overcome deserialization problem. Please help to review.

tdas replied:
LGTM, but I dnt understand the problem well enough to comment with full conviction. I will let @davies check it out.

amit-ramesh commented:
@jerryshao @tdas 
Have a few points from a user perspective regarding the DStream version:

1. Based on this PR it looks like foreachRDD is the only way to get offsets in the DStream case. This would mean that the data needs to be sent over to the driver in order to obtain offsets. Is it possible to obtain offsets in the workers right after receiving the data from Kafka?

2. We have state in updateStateByKey() that necessitates attaching corresponding Kafka metadata to every event in the DStream in order to be able to reconstruct the state across deployments. Is there a way to attach the Kafka offset to every event using the spark API? Essentially obtaining something akin to the output that juanrh had originally proposed in SPARK-8337.

tdas commented:
You can use the alternate version of createDirectStream (returns DStream[T]) that allows you to specify messageHandler function MessageAndMetadata => T . This is applied on every message from Kafka, and you can encode both the message and its offset as a single Spark/RDD record.

BTW, this question unrelated to this PR best asked in the user-mailing list or the relevant JIRA.

amit-ramesh commented:
@tdas
I assume you are talking about the scala API and not python? But maybe it is achievable using DStream.transform() in python?

I asked this here because I originally created SPARK-8337 and wanted to make sure that our use case is addressed :).

jerryshao commented:
@amit-ramesh , the problem you mentioned is different from here what we addressed. Basically I think you required such functionality like `messageHandler` in Scala/Java API to get offset in the executor side. Currently we don't have clear solution to support this in Python API, since Spark Streaming Kafka API on Python is slightly different in implementation compared to Scala/Java, so it is not easy to handle the problem you mentioned unless we change the current semantics.

jerryshao commented:
Even using `DStream.transform()` you could only get the offsetRanges in the driver side, please check the implementation of this PR.

amit-ramesh commented:
@jerryshao
Although the title of SPARK-8337 is worded more broadly, all we really wanted is a way to attach Kafka offsets to individual events/messages :).

AFAICT, the implementation of transform appears to be doing executor side operations. And if the following two assumptions hold then we should be able to achieve what we need:
1. Events in an RDD partition are ordered by Kafka offset
2. The index of an OffsetRanges object in the getOffsets() list corresponds to the partition index in the RDD.

If this is not indeed so, it would be great if you could point me to some lines that use driver side operations in transform.

jerryshao commented:
Hi @amit-ramesh , what I mentioned about getting offsetRanges in `transform` function is something like this:

```python
dstream.transform(lambda r: r.offsetRanges())
```

Here `r.offsetRanges` is executed in driver side, if you have follow-up transformations inside `transfrom` function which need to use this offsetRanges, this offsetRanges will implicitly be sent to executor side. That's what I mean about.

Also:
>1. Events in an RDD partition are ordered by Kafka offset
>2. The index of an OffsetRanges object in the getOffsets() list corresponds to the partition index in the RDD.

This two assumptions are true as I know. so you could rely on this assumptions.

amit-ramesh commented:
@jerryshao 
Thanks for clarifying! I assume only the Kafka metadata moves between the executors and the driver for the rdd.offsetRanges() call while the actual events/messages don't have to go to the driver? Also, is offsetRanges() a driver side operation for scala as well, or is this python specific?

Thanks!

koeninger commented:
Yes, you're understanding things correctly.  Yes, scala works that way as
well.

On Wed, Jul 8, 2015 at 4:20 PM, amit-ramesh <notifications@github.com>
wrote:

> @jerryshao <https://github.com/jerryshao>
> Thanks for clarifying! I assume only the Kafka metadata moves between the
> executors and the driver for the rdd.offsetRanges() call while the actual
> events/messages don't have to go to the driver? Also, is offsetRanges() a
> driver side operation for scala as well, or is this python specific?
>
> Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/7185#issuecomment-119733765>.
>

amit-ramesh commented:
Thanks, @koeninger!

tdas commented on python/pyspark/streaming/util.py:
```diff
@@ -37,6 +37,11 @@ def __init__(self, ctx, func, *deserializers):
         self.deserializers = deserializers
+        self._rdd_wrapper = lambda jrdd, ctx, ser: RDD(jrdd, ctx, ser)
+
+    def rdd_wrapper(self, func):
```
Shouldnt this be called `set_rdd_wrapper`? Can you put some comments on what it does and why it may require setting a custom wrapper?

jerryshao replied:
Hi @tdas , I haven't yet fixed this, I'll change this later on.

tdas commented:
Other than a minor point, it is LGTM.
